### ISSUE TITLE
remove the context from AsyncInfo

### ISSUE THREAD

**Issue by plastikfan** [snivilised/extendio#314]
According to page 133 of ___Concurrency In Go(CiG)___, the context should not be stored on any struct, rather they should always be passed directly into any receiving function.

This means we should pass it into the runner's ___Run___ method and should flow on in a similar fashion from there onwards.

This limitation also extends to cancel functions. These should also never be stored in structs. (I think I have made this mistake somewhere, but it may not even be in extendio, perhaps in ___lorax___, ___pixa___, ___scorpio___ and/or ___arcadia/astrolib___

Consider this interesting chat I had with ___chatgpt___ about context:

---

🎓 User

# What happens if you cancel a context that has already been cancelled in golang

🤖 ChatGPT
In Go, when you call the CancelFunc returned by the context.WithCancel function to cancel a context, it triggers the cancellation of the associated context and all its derived contexts. Once a context is cancelled, all operations that are blocked on that context's Done channel will be unblocked, and they will return a context-related error.

If you attempt to cancel a context that has already been cancelled, nothing specific happens. It's safe to call the CancelFunc multiple times, but once a context is cancelled, subsequent calls to the CancelFunc or any operations dependent on that context's cancellation will not have any further effect. The context remains in a cancelled state, and attempting to use it will result in the same behaviour as if it were cancelled just once.

Here's a simple example to illustrate:

```go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	parentCtx := context.Background()
	ctx, cancel := context.WithCancel(parentCtx)

	// First cancellation
	cancel()

	// Attempt to cancel again (no effect)
	cancel()

	select {
	case <-ctx.Done():
		fmt.Println("Context cancelled")
	case <-time.After(time.Second):
		fmt.Println("Timeout reached")
	}
}
```

In this example, the context is cancelled twice. However, the result will be the same as if it were cancelled just once: the "Context cancelled" message will be printed, and the second cancellation has no additional impact.

Remember that calling cancel() once is enough to cancel the context and all derived contexts, so it's generally not necessary to call it more than once.

---

🎓 User

# Once a cancel function has been obtained from a context function like WithCancel is it mandatory to invoke that cancel function typically in a defer statement

🤖 ChatGPT
It's not mandatory to invoke the cancel function obtained from a context function like context.WithCancel using a defer statement, but it's a good practice to do so in order to ensure that the associated resources are released properly. This helps prevent resource leaks and ensures that your program behaves as expected.

Using defer to call the cancel function ensures that it will be called when the surrounding function returns, regardless of whether it returns normally or due to a panic. This can help you avoid accidental leaks of go-routines or other resources tied to the context.

Here's an example of how you might use defer with a cancel function:

```go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	parentCtx := context.Background()
	ctx, cancel := context.WithCancel(parentCtx)
	defer cancel() // Ensure the cancel function is called when main() returns

	go func() {
		// Simulate some work
		time.Sleep(time.Second)
		fmt.Println("Goroutine completed")
	}()

	select {
	case <-ctx.Done():
		fmt.Println("Context cancelled")
	case <-time.After(2 * time.Second):
		fmt.Println("Timeout reached")
	}
}
```

In this example, the cancel function is deferred immediately after being obtained. This ensures that even if the goroutine takes longer than expected, the cancel function will still be called when main() returns, releasing resources associated with the context.

In summary, while it's not mandatory to use defer with the cancel function, it's a good practice to adopt in order to prevent resource leaks and ensure proper clean-up.

---

🎓 User

# Ok lets not focus on the defer statement. What I really want to know is if we obtain a cancel function but the end user does not cancel their request, do we still need to call the cancel function?

🤖 ChatGPT
If the end user of a context does not explicitly cancel their request and the associated context is not needed any more, it's generally a good practice to call the cancel function to release any resources associated with the context. This helps ensure proper clean-up and can prevent potential resource leaks.

However, if the context is not actively used and the program is about to exit anyway, not calling the cancel function explicitly might not lead to significant issues, as the Go runtime will eventually clean up any leaked resources when the program terminates.

In most cases, it's recommended to have a clear pattern for managing contexts and their cancellations throughout your codebase to ensure proper resource management and prevent unexpected issues. If you're in a situation where a context is no longer needed and the end user has not explicitly cancelled it, calling the cancel function before letting it go out of scope is a responsible approach.

---

🎓 User

# So really the cancel function has been misnamed by go. It is confusing. Normally, one would think that a cancel function would only be used when the end user has cancelled their request. But in reality, the cancel function is also used to clean up resources. Is that correct? If this is correct then the cancel function should really have a different name like dispose?

🤖 ChatGPT
You bring up an interesting point. The name "cancel" function in the context package can indeed be a bit misleading, as it implies that it's primarily for signalling the cancellation of a request. However, its purpose is broader than just that; it's also used to signal the end of the useful lifespan of a context and to clean up resources associated with that context.

The cancel function in the context package serves two main purposes:

Cancellation Signalling: It allows you to signal that a request, operation, or task associated with the context should be cancelled. This is the primary use case that the name "cancel" refers to.

Resource Clean-up: It's used to release resources associated with the context. For example, if you have spawned go-routines, timers, or other resources tied to the context, calling the cancel function ensures these resources are properly cleaned up.

While the name "cancel" might not perfectly encompass both of these aspects, it's important to understand the full range of the function's responsibilities. Renaming it to something like "dispose" could potentially be more descriptive in terms of resource clean-up, but it might not fully capture the cancellation signalling aspect. Finding a name that clearly conveys both meanings while remaining concise can be challenging.

Ultimately, whether it's called "cancel", "dispose", or some other name, the important thing is to understand its purpose and use it responsibly to ensure proper clean-up and resource management in your code.

---

🎓 User

# Ok I have a code base where I am using cancel function, but I am only invoking the cancel function if the client cancels the request. Should I update this to always invoke cancel, even if the user has not cancelled their request.

🤖 ChatGPT
If your codebase currently uses the cancel function only when the client cancels the request and you don't observe any resource leaks or negative consequences, you might not necessarily need to change your approach.

The primary purpose of calling the cancel function is to clean up resources associated with a context. If you're confident that your codebase manages resources properly and you're not encountering issues related to un-canceled contexts, there might be no immediate need to make a change.

However, it's a good practice to ensure that resources are always cleaned up appropriately. In cases where a context is no longer needed and there are resources associated with it (like active go-routines, timers, or other context-related objects), it's safer to call the cancel function explicitly to ensure proper clean-up. This can help prevent potential resource leaks and improve the overall robustness of your code.

If you're concerned about making significant changes to your codebase, you could start by gradually incorporating this practice in new code you write, and over time, you can consider refactoring existing code to follow the same pattern.

Remember that while Go's context package provides the cancel function to signal cancellation and release resources, the specific approach you take will depend on the nature of your application, how contexts are used, and the potential consequences of not properly cleaning up resources.

---

🎓 User

# When you obtain a cancel function, does it have state associated with it that links it with the context?

🤖 ChatGPT
Yes, the cancel function obtained from a context has state associated with it that links it to the context. When you create a new context using functions like context.WithCancel, a new context and a corresponding cancel function are created together, and they share this internal state.

The cancel function operates on this shared internal state to manage the cancellation status of the associated context. When you call the cancel function, it triggers the cancellation of the context and sets its Done channel to a closed state. This closure of the Done channel is what unblocks all operations waiting on the context, indicating that the context has been canceled.

Here's a high-level overview of how the cancel function and context are connected:

Context Creation: When you create a context using context.WithCancel, a new context and its associated cancel function are created. The internal state is managed by the context package.

Cancel Function: The cancel function is a callback that, when called, signals the cancellation of the context. This function is linked to the context and uses the shared internal state to manage the cancellation status.

Cancelling the Context: When the cancel function is called, it updates the internal state of the context to indicate that it has been canceled. It also closes the Done channel associated with the context.

Done Channel: The Done channel is used by operations that need to monitor the cancellation status of the context. When the context is canceled (via the cancel function), the Done channel is closed, and any operations waiting on it are unblocked.

In summary, the cancel function obtained from a context is more than just a simple callback. It's a mechanism that interacts with the context's internal state to signal cancellation and manage associated resources. This design allows Go's context package to provide a clean and effective way to manage the lifecycle and cancellation of operations.

---

🎓 User

# I have read that we should not store the context in any struct, instead we should always pass the context (usually as the first argument) to a function. Does this also apply to the cancel function? Should we never store a cancel function inside a struct?

🤖 ChatGPT
The guidance of not storing a context inside a struct is generally aimed at avoiding the potential misuse or accidental retention of context values beyond their intended lifespan. Contexts are designed to be short-lived and passed explicitly to functions that need them.

The same principles generally apply to the cancel function as well. Storing the cancel function inside a struct can lead to unintended consequences if the struct outlives the intended context's lifespan. When the cancel function is invoked, it affects the associated context and any derived contexts, potentially leading to unexpected behaviour if those contexts are still in use.

Here are a few reasons why it's generally recommended not to store a cancel function inside a struct:

Lifetime Management: The cancel function and the context it is associated with should have the same lifespan. If you store the cancel function in a struct, you need to ensure that the struct is only used while the associated context is still valid.

Avoiding Leaks: If a struct retains a reference to a context's cancel function, it can inadvertently prevent the context from being garbage collected, even if the context is no longer needed. This can lead to resource leaks.

Predictability: Passing the context explicitly to functions makes it clear which context a function is operating on. This helps in maintaining code clarity and predictability.

Instead of storing the cancel function inside a struct, it's recommended to pass the context (and if necessary, the cancel function) as function arguments to the functions that need them. This ensures that the context and cancel function are properly managed and used only in the scope where they are needed.

Remember, there might be certain cases where storing a cancel function in a struct is appropriate, but those cases are likely to be exceptions rather than the norm. Always consider the intended lifespan and usage of the context and cancel function when deciding whether to store them in a struct or pass them explicitly to functions.

---

🎓 User

# Ok so the lifetime of a context should always match the lifetime of the go routine. Is that correct?

🤖 ChatGPT
Yes, you're correct. The lifetime of a context should ideally match the lifetime of the goroutine or operation it is associated with. This ensures that the context and the associated resources are properly managed and cleaned up when they are no longer needed.

When you create a context using functions like context.WithCancel, context.WithTimeout, or context.WithDeadline, it's a good practice to pass this context to the goroutine or operation that you intend to associate with it. The context provides cancellation signals, deadline management, and other functionalities that help you manage the lifecycle of the goroutine.

Here are a few key points to keep in mind regarding the lifetime of a context and its associated goroutine:

Passing Context: Always pass the context explicitly to the goroutine's function or operation as one of the function arguments. This makes it clear which context the goroutine is using and helps manage the lifetime of the context.

Cancelling the Context: If the goroutine finishes its work early or should be terminated due to external conditions, you can call the context's cancel function to signal the cancellation. This can help the goroutine clean up resources gracefully.

Context Hierarchies: When you create derived contexts using functions like context.WithValue or context.WithTimeout, ensure that their lifetimes match the corresponding go-routines or operations. Derived contexts inherit the cancellation signal from their parent contexts.

Avoiding Leaks: To prevent resource leaks, make sure that the context is not kept alive beyond its intended lifespan. If a goroutine or operation completes and you no longer need the context, let the context fall out of scope so it can be garbage collected.

By ensuring that the context's lifetime matches the goroutine's lifetime, you can effectively manage resources, handle cancellations, and maintain the integrity of your program's design.

---

🎓 User

# If a parent context is cancelled, does that mean the child context is also cancelled, but when a child context is cancelled this does not mean the parent context is cancelled. Is this correct?

🤖 ChatGPT
Yes, you're correct. When a parent context is cancelled, it will propagate the cancellation signal to all its derived child contexts. However, if a child context is cancelled, it does not automatically cancel the parent context.

This behaviour ensures that you can manage the cancellation of different parts of your program independently, while still allowing for cascading cancellations when higher-level contexts are cancelled.

Here's a summary of how context cancellation propagation works:

Parent Context Cancellation: When you call the cancel function on a parent context, it cancels not only that context but also all derived child contexts created from it. Any operations blocked on the Done channel of those child contexts will be unblocked, and they will receive a signal that the context has been cancelled.

Child Context Cancellation: If you cancel a child context, it does not automatically cancel the parent context. The parent context and other sibling child contexts will continue to function as normal unless explicitly cancelled.

This design allows you to manage different parts of your program independently and ensures that you can cancel specific operations or parts of the program without affecting the entire program. It's important to note that each context can have its own cancellation signal, timeout, and other attributes, allowing for fine-grained control over the behaviour of your concurrent operations.
```

**Comment by plastikfan** [comment]
This issue is proving too difficult to fix, without a major refactor of the front end (the session and runner). Not storing the context as a member of a struct really complicates thing as we also need to pass in the parent cancel function, so we can't use a variadic parameter as we have 2 different types that need to be passed in. The only way we can do this is to ___interface{}___, but we then lose type safety.

This refactor will have to take place in a different issue. This issue will implement some minor refactorings including moving the deferred job channel closure and wait group done signalling to the accelerator.
